### ISSUE TITLE
Stack overflow at compile time when using persist-mongodb

### ISSUE THREAD

**Issue by jsantos17** [yesodweb/persistent#311]
So I've been trying to update to persistent-2.1 from persistent-1.3 and I'm getting a stack overflow during compilation in the following Template Haskell:

```
let mongoSettings = (mkPersistSettings (ConT ''MongoDB.MongoContext))
                        { mpsGeneric = False }
  in share [mkPersist mongoSettings]
     $(persistFileWith lowerCaseSettings "modelsMongo")
```

I got the previous example directly from persistent-tests and made it use my own model definition file (i.e modelsMongo). I think I might be overlooking something simple. Did the syntax for defining models change or something like that?

Thanks in advance!

**Comment by gregwebs** [comment]
In the switch from persistent 1 to 2 I improved the types that describe entities. But that created the potential for an infinite structure. Haskell can handle infinite structures. However, lifting the structure through the Q monad fully evaluates the structure. I created a pull request to break the cycle: #372

Thanks for your patience.

**Comment by gregwebs** [comment]
Fixed in persistent-template-2.1.3

**Comment by obscaenvs** [comment]
Great, thx, @gregwebs!
